source/common/config/type_to_endpoint.cc

```diff
@@ -6,96 +6,170 @@
 
 // API_NO_BOOST_FILE
 
+#define API_VERSION envoy::config::core::v3::ApiVersion
+#define VERSIONED(v2, v3)                                                                          \
+  createServiceToVersionedMap(v2, API_VERSION::V2, {v2, v3}),                                      \
+      createServiceToVersionedMap(v3, API_VERSION::V3, {v2, v3})
+
 namespace Envoy {
 namespace Config {
 
 namespace {
 
-// service RPC method fully qualified names.
-struct Service {
-  std::string sotw_grpc_method_;
-  std::string delta_grpc_method_;
-  std::string rest_method_;
+struct VersionedMap {
+  envoy::config::core::v3::ApiVersion version_;
+  std::unordered_map<envoy::config::core::v3::ApiVersion, std::string> entries_;
+};
+
+// Map from resource type URL to versioned service RPC methods.
+using ServiceToVersionedMap = std::unordered_map<std::string, VersionedMap>;
```

Comment:
Is all this complexity because you want to support resource/transport independence still? I think it might be clearer if you provide some example maps and explain how this structure is used.

Reply:
Haha. I think I made this too complicated in the price to support `AUTO` to default to the inferred "version" from a particular resource type URL.